pynbody/snapshot/__init__.py

```diff
@@ -35,8 +35,105 @@
 
 logger = logging.getLogger('pynbody.snapshot')
 
+class UnitFullContainer:
+    """
+    Defines an abstract class that has properties and arrays
+    that can be converted to physical units.
+    """
+    _autoconvert = None
+
+    def _get_dims(self, dims=None):
+        if dims is None:
+            return self.ancestor._autoconvert
+        else:
+            return dims
+
+
+    def _autoconvert_array_unit(self, ar, dims=None, ucut=3):
+        """Given an array ar, convert its units such that the new units span
+        dims[:ucut]. dims[ucut:] are evaluated in the conversion (so will be things like
+        a, h etc).
+
+        If dims is None, use the internal autoconvert state to perform the conversion."""
+        dims = self._get_dims(dims)
+        if dims is None:
+            return
+
+        if ar.units is not None:
+            try:
+                d = ar.units.dimensional_project(dims)
+            except units.UnitsException:
+                return
+
+            new_unit = reduce(lambda x, y: x * y, [
+                              a ** b for a, b in zip(dims, d[:ucut])])
+            if new_unit != ar.units:
+                logger.info("Converting %s units from %s to %s" %
+                            (ar.name, ar.units, new_unit))
+                ar.convert_units(new_unit)
+
+
+    def _autoconvert_properties(self, dims=None):
+        dims = self._get_dims(dims)
+        if dims is None:
+            return
+
+        for k, v in list(self.properties.items()):
+            if isinstance(v, units.UnitBase):
+                try:
+                    new_unit = v.dimensional_project(dims)
+                except units.UnitsException:
+                    continue
+                new_unit = reduce(lambda x, y: x * y, [
+                                  a ** b for a, b in zip(dims, new_unit[:3])])
+                new_unit *= v.ratio(new_unit, **self.conversion_context())
+                self.properties[k] = new_unit
+            elif isinstance(v, array.SimArray):
+                self._autoconvert_array_unit(v, dims)
+
+    def _autoconvert_arrays(self, dims=None):
+        dims = self._get_dims(dims)
+        if dims is None:
+            return
+
+        all = list(self._arrays.values())
+        for x in self._family_arrays:
+            all += list(self._family_arrays[x].values())
+
+        for ar in all:
+            self._autoconvert_array_unit(ar.ancestor, dims)
+
+    def physical_units(self, distance='kpc', velocity='km s^-1', mass='Msol', persistent=True):
```

Comment:
Should you add and document the `convert_parent` kwarg here, to save potential confusion? It can simply have no effect if there is no parent object.

Reply:
I added docstrings to `physical_units` with a different explanation for `persistent` depending on context. Tell me what you think